pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,217 @@
+	User      string // FTP username
+	Password  string // FTP password
+	Port      string // FTP port
+	RemoteDir string // Remote directory path
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) datasource.FileSystem {
+	return &ftpFileSystem{config: config}
+}
+
+// UseLogger sets the logger interface for the FTP file system.
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	f.logger = logger.(datasource.Logger)
+}
+
+// UseMetrics sets the metrics for the ftpFileSystem client which asserts the Metrics interface.
+// Currently not implemented.
+func (*ftpFileSystem) UseMetrics(metrics interface{}) {
+
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	// Construct FTP server address
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	// Connect to FTP server
+	const dialTimeoutSeconds = 5
+
+	conn, err := ftp.Dial(ftpServer, ftp.DialWithTimeout(dialTimeoutSeconds*time.Second))
+	if err != nil {
+		log.Printf("Failed to connect to FTP server: %v", err)
+		return
+	}
+
+	f.conn = &Conn{conn}
+
+	// Login to FTP server
+	err = conn.Login(f.config.User, f.config.Password)
+	if err != nil {
+		log.Printf("Failed to login: %v", err)
+	} else {
+		log.Printf("Login Successful")
+	}
+}
+
+// Create creates an empty file on the FTP server.
+func (f *ftpFileSystem) Create(name string) (datasource.File, error) {
+	// empty io.Reader
+	emptyReader := new(bytes.Buffer)
+
+	if name == "" {
+		return nil, errors.New("empty filename")
+	}
+
+	// construct the path
+	name = fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	// Issue STOR command to create the empty file on FTP server
+	err := f.conn.Stor(name, emptyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Created file %s", name)
+
+	s := strings.Split(name, "/")
+
+	// Retrieve the file from FTP server and return a file handle
+	res, err := f.conn.Retr(name)
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     name,
+		conn:     f.conn,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	if name == "" {
+		return errors.New("empty directory name")
+	}
+
+	name = fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	return f.conn.MakeDir(name)
+}
+
+// MkdirAll creates directories recursively on the FTP server.
+func (f *ftpFileSystem) MkdirAll(path string, _ os.FileMode) error {
+	// Split path into individual directory names
+	dirs := strings.Split(path, "/")
+
+	// Start creating directories
+	currentDir := ""
+
+	for _, dir := range dirs {
+		// Ignore empty directory names (can happen if there are double slashes)
+		if dir == "" {
+			continue
+		}
+
+		// Build the full path up to this directory
+		currentDir = filepath.Join(currentDir, dir)
```

Comment:
Do not use `filepath` for building remote paths. A windows machine running this code would generate the path: `"C:\\Windows"` which the remote service would not understand.

Reply:
Noted.